### PR TITLE
Remove Dependency between Bart and LED (slow/fast)

### PR DESCRIPTION
**Pull request by Infrared1029**
# What does this PR do?

Removes the dependency between LED and Bart

## Who can review?
@sgugger

### CODE REVIEW

**Comment by Infrared1029** [comment]
@sgugger hopefully this does it?

File: src/transformers/models/led/tokenization_led.py [hunk 1]
```diff
@@ -40,18 +49,365 @@
...
+    """
+    pairs = set()
+    prev_char = word[0]
+    for char in word[1:]:
+        pairs.add((prev_char, char))
+        prev_char = char
+    return pairs
+
+
+# Copied from transformers.models.bart.tokenization_bart.BartTokenizer with Bart->LED,facebook/bart-base->allenai/led-base-16384
```
Comment (sgugger) [inline src/transformers/models/led/tokenization_led.py:93]:
```suggestion
# Copied from transformers.models.bart.tokenization_bart.BartTokenizer with Bart->LED,BART-LED,facebook/bart-base->allenai/led-base-16384
```

Comment (Infrared1029) [reply src/transformers/models/led/tokenization_led.py:93]: Whoops, on it

**Comment by sgugger** [comment]
Just tried locally your PR and found the reason why we can't have a global copied from on the tokenizer (sorry I didn't spot it earlier): the `_pad` method is overwritten.
So instead we need to applied the copied from on each method (except `_pad`) and not copy the whole class.

Sorry about that!

**Comment by Infrared1029** [comment]
> Just tried locally your PR and found the reason why we can't have a global copied from on the tokenizer (sorry I didn't spot it earlier): the `_pad` method is overwritten. So instead we need to applied the copied from on each method (except `_pad`) and not copy the whole class.
> 
> Sorry about that!

so basically, write those copy comments again on both the slow and fast tokenizers?

**Comment by sgugger** [comment]
Yeah, sorry

**Comment by Infrared1029** [comment]
> Yeah, sorry

Oh dont be, on it now :D, thanks for the quick reply!

**Comment by Infrared1029** [comment]
@sgugger anything left?

File: src/transformers/models/led/tokenization_led.py [hunk 2]
```diff
@@ -40,18 +49,378 @@
...
+    """
+    pairs = set()
+    prev_char = word[0]
+    for char in word[1:]:
+        pairs.add((prev_char, char))
+        prev_char = char
+    return pairs
+
+
+# Copied from transformers.models.bart.tokenization_bart.BartTokenizer with Bart->LED,BART-LED,facebook/bart-base->allenai/led-base-16384
```
Comment (sgugger) [inline src/transformers/models/led/tokenization_led.py:93]:
```suggestion
```
This one needs to be removed (or it overwrites all your other changes :-) )

File: src/transformers/models/led/tokenization_led_fast.py [hunk 3]
```diff
@@ -41,19 +47,251 @@
 }
 
 
-class LEDTokenizerFast(BartTokenizerFast):
+# Copied from transformers.models.bart.tokenization_bart_fast.BartTokenizerFast with Bart->LED,BART-LED,facebook/bart-base->allenai/led-base-1638
```
Comment (sgugger) [inline src/transformers/models/led/tokenization_led_fast.py:50]:
```suggestion
```
Same here

File: src/transformers/models/led/tokenization_led.py (hunk 2 above)
Comment (Infrared1029) [reply src/transformers/models/led/tokenization_led.py:93]: oh right we don't need this anymoreD:

**Comment by Infrared1029** [comment]
btw, the `mask_token` method in the fast tokenizer, there is a method and a setter , but i have the same comment on both (Copied from `BartTokenizerFast.mask_token)` is that the right way or am i supposed to keep the one above the method only and ignore the setter's one?

File: src/transformers/models/led/tokenization_led.py [hunk 4]
```diff
@@ -40,18 +49,377 @@
...
+                        f"Saving vocabulary to {merge_file}: BPE merge indices are not consecutive."
+                        " Please check that the tokenizer is not corrupted!"
+                    )
+                    index = token_index
+                writer.write(" ".join(bpe_tokens) + "\n")
+                index += 1
+
+        return vocab_file, merge_file
+
+    # Copied from transformers.models.bart.tokenization_bart.BartTokenizer.build_inputs_with_special_tokens
```
Comment (sgugger) [inline src/transformers/models/led/tokenization_led.py:339]:
```suggestion
    # Copied from transformers.models.bart.tokenization_bart.BartTokenizer.build_inputs_with_special_tokens with BART->LED
```

File: src/transformers/models/led/tokenization_led.py [hunk 5]
```diff
@@ -40,18 +49,377 @@
...
+        if already_has_special_tokens:
+            return super().get_special_tokens_mask(
+                token_ids_0=token_ids_0, token_ids_1=token_ids_1, already_has_special_tokens=True
+            )
+
+        if token_ids_1 is None:
+            return [1] + ([0] * len(token_ids_0)) + [1]
+        return [1] + ([0] * len(token_ids_0)) + [1, 1] + ([0] * len(token_ids_1)) + [1]
+
+    # Copied from transformers.models.bart.tokenization_bart.BartTokenizer.create_token_type_ids_from_sequences
```
Comment (sgugger) [inline src/transformers/models/led/tokenization_led.py:393]:
```suggestion
    # Copied from transformers.models.bart.tokenization_bart.BartTokenizer.create_token_type_ids_from_sequences with BART->LED
```

File: src/transformers/models/led/tokenization_led_fast.py [hunk 6]
```diff
@@ -41,19 +47,250 @@
...
+        comprise the space before the *<mask>*.
+        """
+        if self._mask_token is None:
+            if self.verbose:
+                logger.error("Using mask_token, but it is not set yet.")
+            return None
+        return str(self._mask_token)
+
+    @mask_token.setter
+    # Copied from transformers.models.bart.tokenization_bart_fast.BartTokenizerFast.mask_token
```
Comment (sgugger) [inline src/transformers/models/led/tokenization_led_fast.py:222]:
```suggestion
```

File: src/transformers/models/led/tokenization_led_fast.py [hunk 7]
```diff
@@ -41,19 +47,250 @@
...
+                state["trim_offsets"] = trim_offsets
+                changes_to_apply = True
+
+            if changes_to_apply:
+                component_class = getattr(processors, state.pop("type"))
+                new_value = component_class(**state)
+                setattr(self.backend_tokenizer, tokenizer_component, new_value)
+
+    @property
+    # Copied from transformers.models.bart.tokenization_bart_fast.BartTokenizerFast.mask_token
```
Comment (sgugger) [inline src/transformers/models/led/tokenization_led_fast.py:206]:
```suggestion
    # Copied from transformers.models.bart.tokenization_bart_fast.BartTokenizerFast.mask_token with BART->LED
```

File: src/transformers/models/led/tokenization_led_fast.py [hunk 8]
```diff
@@ -41,19 +47,250 @@
...
+
+    # Copied from transformers.models.bart.tokenization_bart_fast.BartTokenizerFast.build_inputs_with_special_tokens
+    def build_inputs_with_special_tokens(self, token_ids_0, token_ids_1=None):
+        output = [self.bos_token_id] + token_ids_0 + [self.eos_token_id]
+        if token_ids_1 is None:
+            return output
+
+        return output + [self.eos_token_id] + token_ids_1 + [self.eos_token_id]
+
+    # Copied from transformers.models.bart.tokenization_bart_fast.BartTokenizerFast.create_token_type_ids_from_sequences
```
Comment (sgugger) [inline src/transformers/models/led/tokenization_led_fast.py:271]:
```suggestion
    # Copied from transformers.models.bart.tokenization_bart_fast.BartTokenizerFast.create_token_type_ids_from_sequences with BART->LED
```

**Review by sgugger — COMMENTED** [review]
Indeed, you can't apply the copied from on the setter as it tries to copy the getter code. So let's remove that copied from statement.

Otherwise, just detected a couple of instances where there is a BART to replace to LED in the docstrings, it's all in the suggestions below.

**Comment by Infrared1029** [comment]
thanks a lot for the quick replies, what else is left?

**Comment by sgugger** [comment]
Should be good now, just waiting for all tests to pass :-)

**Comment by Infrared1029** [comment]
oh finally, the green light

**Review by sgugger — APPROVED** [review]
Thanks again for your contribution!

**Pull request merged**